tensorflow/core/util/mkl_util.h

```diff
@@ -2061,6 +2060,99 @@ class MklPrimitive {
 
 const mkldnn::memory::dims NONE_DIMS = {};
 
+//
+// LRUCache is a class which implements LRU (Least Recently Used) cache.
+// The implementation is similar to that of
+//    tensorflow/core/platform/cloud/expiring_lru_cache.h
+// without its thread-safe part because the cache is supposed to be
+// used as thread local (for instance, MklPrimitive caching).
+//
+// The LRU list maintains objects in chronological order based on
+// creation time, with the least recently accessed object at the
+// tail of LRU list, while the most recently accessed object
+// at the head of LRU list.
+//
+// This class is used to maintain an upper bound on the total number of
+// cached items. When the cache reaches its capacity, the LRU item will
+// be removed and replaced by a new one from SetOp call.
+//
+template <typename T>
+class LRUCache {
+ public:
+  explicit LRUCache(size_t capacity) {
+    capacity_ = capacity;
+    Clear();
+  }
+
+  T* GetOp(const string& key) {
+    auto it = cache_.find(key);
+    if (it == cache_.end()) {
+      return nullptr;
+    }
+
+    // Move to the front of LRU list as the most recently accessed.
+    lru_list_.erase(it->second.lru_iterator);
+    lru_list_.push_front(it->first);
+    it->second.lru_iterator = lru_list_.begin();
+    return it->second.op;
+  }
+
+  void SetOp(const string& key, T* op) {
+    if (lru_list_.size() >= capacity_) {
+      Delete();
+    }
+
+    // Insert an entry to the front of the LRU list
+    lru_list_.push_front(key);
+    Entry entry(op, lru_list_.begin());
+    cache_.insert(std::make_pair(key, entry));
+  }
+
+  void Clear() {
+    if (lru_list_.empty()) return;
+
+    // clean up the cache
+    cache_.clear();
+    lru_list_.clear();
+  }
+
+ private:
+  struct Entry {
+    // The entry's value.
+    T* op;
+
+    // A list iterator pointing to the entry's position in the LRU list.
+    std::list<string>::iterator lru_iterator;
+    Entry(T* op, std::list<string>::iterator it) {
+      this->op = op;
+      this->lru_iterator = it;
+    }
```

Comment:
Please add a [move constructor](https://en.cppreference.com/w/cpp/language/move_constructor) and a destructor.
```c++
Entry(Entry&& source) noexcept
    : op(std::exchange(source.op, nullptr)), it(std::move(source.it)) {}
~Entry() {
  if(source.op != nullptr) delete source.op;  
}
```

Reply:
Penporn, 
Thank you very much for the help. I just committed the changes based on your ideas.
Please note the small tweak in my implementation
 1. Do not use std::exchange method as it is for C++ 14 while we need to support C++ 11 in TF.
 2. Minor change in the destructor, as "source is not needed" there. 
Best,
GZ

Reply:
Thank you for your patience! 
1: Sorry! I wasn't aware of that. I'll keep that in mind. :)
2: Oops. Yes.